Fish 10526 improving concurrent implementation of corba to use lock api

breakponchito opened the pull request:
This is to improve the current implementation to instead of using synchronized blocks to prevent concurrent threads to access resources, now use lock api. We decided to improve because with synchronized it is hard to determine which thread access and release the resources. Here I list the advantages to use the lock api:

1. Fine-Grained control: We are doing explicit locking and unlocking, this is more precise and efficient than the implementation with synchronized
2. Use of conditions variables: This allow threads to wait for certain conditions to be met before proceeding
3. Multiple Object locking: lock can be shared among multiple class instances, allowing you to manage synchronization across
these instances more flexible

For the tests I created the following Remote EJB:

[remoteview.zip](https://github.com/user-attachments/files/18833979/remoteview.zip)

To use you need to deploy on the server and include this jar on an standalone java application, here included also:
[EJBRemoteClient5.zip](https://github.com/user-attachments/files/18833994/EJBRemoteClient5.zip)

On the Main class you will see the remote call configuration to execute the tests with multiple threads creating 300 requests to the server. Something important for the standalone application is to indicate the correct path for the modules folder and resolve all dependencies for the EJB Client call configuration:

![image](https://github.com/user-attachments/assets/54683c83-890a-428b-874e-f24e7077d019)

to enable the new implementation with locks you need to set the following property on the system:

`com.sun.corba.ee.protocol.enablingNewFragmentProcess=true`

by default this is set as false to use the synchonized mechanism

also you can set the timeout time of the internal management of the process to wait for information on the queue with the following property:

`com.sun.corba.ee.protocol.newFragmentEmptyConditionTimeout=10000`

by default this is set with 10000 milliseconds, the unit for the value is milliseconds and this value is only used when the previous property is set as true

after doing multiple calls using the reproducer now I can't see any affectation on the threads, now all of them are finishing and also the performance is not affected:

![image](https://github.com/user-attachments/assets/4b51144c-8812-4228-b076-e60bb588c8aa)


I used the following environment for testing: Payara 5, azul JDK 8, windows 11 and maven 3.8.6

breakponchito committed:
FISH-10526: adding logs to identify deadlock

breakponchito committed:
FISH-10526: adding more thread information to identify deadlock

breakponchito committed:
FISH-10526: adding requestId for fragment processing on logs

breakponchito committed:
FISH-10526: just reordering message on logs

breakponchito committed:
FISH-10526: removing queue reference from logs to not cause ConcurrentModificationException

breakponchito committed:
FISH-10526: improving code to use lock api instead of synchronized block

breakponchito committed:
FISH-10526: adding property for timeout and code improvements based on comments

breakponchito committed:
FISH-10526: updating version for next release

Pandrex247 commented on orbmain/src/main/java/com/sun/corba/ee/impl/protocol/MessageMediatorImpl.java:
```diff
@@ -114,18 +119,19 @@
+    private static final Logger logger = Logger.getLogger(MessageMediatorImpl.class.getName());
+    protected static final ORBUtilSystemException wrapper = ORBUtilSystemException.self;
+    protected static final InterceptorsSystemException interceptorWrapper = InterceptorsSystemException.self;
+    private static final String ENABLING_NEW_FRAGMENT_CONCURRENCY_PROCESS = "fish.payara.corba.protocol.enablingNewFragmentProcess";
```
I'm a little wary of introducing payara properties to an external GlassFish project.
These should either follow the same name convention as other properties in Corba, or be reworked and something configured on the Payara side which we pass over to Corba

Pandrex247 commented on orbmain/src/main/java/com/sun/corba/ee/impl/protocol/MessageMediatorImpl.java:
```diff
@@ -730,85 +726,203 @@ private void connectionInfo( Connection conn ) { }
-            // it will be the thread that executes the Work (messageMediator)we
-            // put the on the WorkQueue here.
-            addMessageMediatorToWorkQueue(messageMediator);
+        if (message.moreFragmentsToFollow() && !isNewFragmentProcessingSet) {
+            synchronizedProcess(message);
+        } else if (message.moreFragmentsToFollow() && isNewFragmentProcessingSet) {
+            lockProcess(message);
         } else {
```
Could be cleaned up slightly

```suggestion
        if (message.moreFragmentsToFollow()) {
            if (!isNewFragmentProcessingSet) {
                synchronizedProcess(message);
            } else {
                lockProcess(message);
            }
        } else {
```

breakponchito replied:
already done

Pandrex247 commented on orbmain/src/main/java/com/sun/corba/ee/impl/protocol/MessageMediatorImpl.java:
```diff
@@ -730,85 +726,203 @@ private void connectionInfo( Connection conn ) { }
+                if (queue.size() > 0) {
+                    messageMediator = queue.poll();
+                } else {
+                    if(!queueStillEmpty){
```
```suggestion
                    if (!queueStillEmpty) {
```

breakponchito replied:
already done

breakponchito committed:
FISH-10526: attending comments from PR

Pandrex247 reviewed (dismissed):
The system properties should probably be moved to the `ORBConstants` class

breakponchito committed:
FISH-10526: more fixes from comments on PR

breakponchito committed:
FISH-10526: adding missed copyright

breakponchito commented:
> The system properties should probably be moved to the `ORBConstants` class

already done

Pandrex247 commented on orbmain/src/main/java/com/sun/corba/ee/spi/misc/ORBConstants.java:
```diff
@@ -637,6 +639,10 @@ public static int makePersistent( int scid )
         + "ORBGmbalRootParentName" ;
+    
+    //protocol constants to enable new lock api mechanism to process fragments 
+    public static final String ENABLING_NEW_FRAGMENT_CONCURRENCY_PROCESS = SUN_PREFIX + "protocol.enablingNewFragmentProcess";
```
Final niggle - 'enable' makes more sense here than 'enabling'

```suggestion
    public static final String ENABLE_NEW_FRAGMENT_CONCURRENCY_PROCESS = SUN_PREFIX + "protocol.enableNewFragmentProcess";
```

Pandrex247 committed:
Adjust property name

'Enable' makes more sense than 'enabling'

Pandrex247 reviewed (changes requested):
Missing release commit

Pandrex247 commented:
https://payara.atlassian.net/wiki/spaces/PAYAR/pages/150601767/Patching+External+Projects+Process

Pandrex247 committed:
Adjust property name usage

Missed these when renaming the property name in 2c7c9377492b8d582141e2c0ff8f9e6a7b2d5ed4

Signed-off-by: Andrew Pielage <pandrex247@hotmail.com>

Pandrex247 committed:
Revert "FISH-10526: updating version for next release"

This reverts commit 1fc8490d4e7ff4044705a135ec8e71a984838403.

Pandrex247 committed:
glassfish-corba-4.1.1.payara-p9 release

Pandrex247 committed:
Increment version for development
